utils/emmylua/wesnoth.lua

```diff
@@ -0,0 +1,272 @@
+---@field event_context event_context
+wesnoth.current = {}
+
+---Holds global game configuration options
+---@class game_config
+---@field base_income integer
+---@field village_income integer
+---@field village_support integer
+---@field poison_amount integer
+---@field rest_heal_amount integer
+---@field recall_cost integer
+---@field combat_experience integer
+---@field kill_experience integer
+---@field global_traits table<string, WMLTable>
+---@field do_healing boolean
+---@field theme string
+---@field debug boolean
+---@field debug_lua boolean
+---@field mp_debug boolean
+---@field strict_lua boolean
+wesnoth.game_config = {}
+
+---@type table<string, unit_race>
+wesnoth.races = {}
+
+---Holds information about the current scenario
+---@class scenario
+---@field turns integer
+---@field next string|nil
+---@field id string
+---@field defeat_music string[]
+---@field victory_music string[]
+---@field show_credits boolean
+---@field end_text tstring
+---@field end_text_duration number
+---@field difficulty string
+---@field type "'scenario'"|"'test'"|"'multiplayer'"
+---@field era WMLTable|nil
+---@field campaign WMLTable|nil
+---@field resources WMLTable[]
+---@field modifications WMLTable[]
+---@field mp_settings mp_settings
+wesnoth.scenario = {}
+
+---@class mp_settings
+---@field game_name string
+---@field hash string
+---@field experience_modifier integer
+---@field mp_countdown boolean
+---@field mp_countdown_init_time integer
+---@field mp_countdown_turn_bonus integer
+---@field mp_countdown_reservoir_bonus integer
+---@field mp_countdown_action_bonus integer
+---@field mp_village_gold integer
+---@field mp_village_support integer
+---@field mp_fog boolean
+---@field mp_shroud boolean
+---@field mp_use_map_settings boolean
+---@field mp_random_start_time boolean
+---@field allow_observers boolean
+---@field private_replay boolean
+---@field shuffle_sides boolean
+---@field random_faction_mode "'Independent'"|"'No Mirror'"|"'No Ally Mirror'"
+---@field options WMLTable
+---@field savegame 'false'|"'midgame'"|"'scenario_start'"
+---@field side_players table<string, string>
+---@field addons mp_addon[]
+---@class mp_addon
+---@field id string
+---@field name string
+---@field required boolean
+---@field min_version? string
+---@field version? string
+---@field content {id:string, name:string, type:string}
+
+---Information on a specific terrain type
+---@class terrain_info
+---@field id string
+---@field name tstring
+---@field editor_name tstring
+---@field description tstring
+---@field icon string
+---@field editor_image string
+---@field light integer
+---@field village boolean
+---@field castle boolean
+---@field keep boolean
+---@field healing boolean
+
+---@type table<string, terrain_info>
+wesnoth.terrain_types = {}
+
+---@type table<string, unit_type>
+wesnoth.unit_types = {}
+
+---@generic T
+---@param values T[]
+---@param names string[]
+---@return table<string, T>
+function wesnoth.named_tuple(values, names) end
```

Comment:
Can values be a heterogeneous array? Using `T` here suggest it's homogeneous.

Reply:
That's a good point. I think _most_ of the actual uses are indeed homogenous, but the API itself doesn't care about the content of the array. I'm not sure there's a good way to represent that in the documentation though.

Reply:
A quick search for `emmylua heterogeneous` suggests no-one's asked that yet, so accepted.